### PR TITLE
Feat/92 - Add details of officials to calendar files

### LINKED ISSUE #92: [FEATURE REQUEST]: Add appointed umpires/officials to the match descriptions
### Explain your feature.

Add the details of appointed officials from Altius to the match descriptions

### Added value.

Initial feature to allow for personal calendar subscription

### Additional info

Self developed- Draft PR coming!

### PR DESCRIPTION
**Pull request by philipohagan**
Closes #92

Replaces #93 

## Proposed Changes
- Add the ability (optionally) for the TMS fetcher to request Officals
- Scrape Officials from Altius
- Place the Officials for each match into the appropriate ICS files

## Checklist

- [x] I have read the [contribution document](https://github.com/Martijn-van-Kekem-Development/hockey-match-calendar/blob/main/CONTRIBUTING.md).
- [x] I have added test cases to cover my changes (if applicable).
- [x] All new and existing test cases have passed.
- [x] My code adheres to the coding style of this project.

### CODE REVIEW

**Comment by philipohagan** [comment]
@martijnvankekem - no rush on reviewing but I've had another bash at implementing using your guidance- not accomplished everything but it works well from my testing.

File: src/Fetchers/TMSFetcher/TMSFetcher.ts [hunk 1]
```diff
@@ -75,41 +75,50 @@ export class TMSFetcher extends Fetcher {
...
-        for (const competition of competitions.values()) {
-            // Fetch match for every competition
-            const matchPromise = this.fetchMatches(competition);
-            matchPromise.then(result => {
-                competition.getMatches().push(...result.values());
-                return ICSCreator.createCompetitionICS(competition);
-            });
-
-            promises.push(matchPromise);
+        try {
```
Comment (martijnvankekem) [inline src/Fetchers/TMSFetcher/TMSFetcher.ts:78]: Why is the try-catch necessary here? Which errors are we possibly trying to catch? I'd like a better solution for this than simply catching all errors and logging them, because this will not fail the auto-update script if something goes wrong.

File: src/Fetchers/KNHBFetcher/KNHBClubFetcher.ts [hunk 2]
```diff
@@ -36,14 +36,17 @@ export class KNHBClubFetcher {
...
+     * @param input The input string.
      */
-    public static simplifyString(input: string) {
+    public static simplifyString(input: string | null): string {
+        if (!input) return "";
+
         return input
             .toLowerCase()
-            .replaceAll(/[^a-z]/g, "");
+            .normalize("NFD")
```
Comment (martijnvankekem) [inline src/Fetchers/KNHBFetcher/KNHBClubFetcher.ts:47]: Why is this necessary to implement officals?

File: src/Fetchers/KNHBFetcher/KNHBMatchFetcher.ts [hunk 3]
```diff
@@ -139,16 +139,12 @@ export class KNHBMatchFetcher {
...
-        const homeTeam = matchToParse.home_team?.club_name ??
-            matchToParse.home_team?.name;
-        const awayTeam = matchToParse.away_team?.club_name ??
-            matchToParse.away_team?.name;
-
-        const homeKNHBClub = clubs.get(
-            KNHBClubFetcher.simplifyString(homeTeam));
-        const awayKNHBClub = clubs.get(
-            KNHBClubFetcher.simplifyString(awayTeam));
+        const homeKNHBClub = matchToParse.home_team?.club_name
```
Comment (martijnvankekem) [inline src/Fetchers/KNHBFetcher/KNHBMatchFetcher.ts:142]: Why do we need to simplify the string for this PR?

File: src/Fetchers/TMSFetcher/TMSFetcher.ts [hunk 4]
```diff
@@ -139,9 +148,39 @@ export class TMSFetcher extends Fetcher {
...
+            const roleOrder = [
+                "Umpire",
+                "Reserve/Video",
+                "Scoring/Timing",
+                "Technical Officer"
+            ];
+
+            // Sort officials by predefined role order
+            const sortedOfficials = [...officials].sort((a, b) => {
+                const indexA = roleOrder.indexOf(a.role);
```
Comment (martijnvankekem) [inline src/Fetchers/TMSFetcher/TMSFetcher.ts:169]: This needs a `'?? 0'` in case we encounter a role not in the fixed order list.

**Review by martijnvankekem — COMMENTED** [review]
There are quite some changes to the KNHB fetchers, while there is no actual implementation there. See line-specific comments for more details.

Why are those necessary?

File: src/Fetchers/KNHBFetcher/KNHBClubFetcher.ts (hunk 2 above)
Comment (philipohagan) [reply src/Fetchers/KNHBFetcher/KNHBClubFetcher.ts]: It was solving a bug at some point but should have been removed.

File: src/Fetchers/KNHBFetcher/KNHBMatchFetcher.ts (hunk 3 above)
Comment (philipohagan) [reply src/Fetchers/KNHBFetcher/KNHBMatchFetcher.ts]: We don't - removed.

File: src/Fetchers/TMSFetcher/TMSFetcher.ts (hunk 4 above)
Comment (philipohagan) [reply src/Fetchers/TMSFetcher/TMSFetcher.ts:169]: Done

File: src/Fetchers/TMSFetcher/TMSFetcher.ts (hunk 1 above)
Comment (philipohagan) [reply src/Fetchers/TMSFetcher/TMSFetcher.ts:78]: Relic from testing, my bad.

File: src/Fetchers/TMSFetcher/TMSFetcher.ts [hunk 5]
```diff
@@ -75,41 +75,45 @@ export class TMSFetcher extends Fetcher {
      * Fetch the matches from TMS.
      */
     protected async fetch() {
```
Comment (martijnvankekem) [inline src/Fetchers/TMSFetcher/TMSFetcher.ts:77]: It doesn't seem like there are any changes to this function, except for a try-catch solution instead of promise-based. Please revert this to the original implementation, as it's more readable and does functionally the same.

File: src/Fetchers/TMSFetcher/TMSFetcher.ts [hunk 6]
```diff
@@ -139,9 +143,39 @@ export class TMSFetcher extends Fetcher {
...
+            const roleOrder = [
+                "Umpire",
+                "Reserve/Video",
+                "Scoring/Timing",
+                "Technical Officer"
+            ];
+
+            // Sort officials by predefined role order
+            const sortedOfficials = [...officials].sort((a, b) => {
+                const indexA = roleOrder.indexOf(a.role) ?? 0;
```
Comment (martijnvankekem) [inline src/Fetchers/TMSFetcher/TMSFetcher.ts:164]: indexOf takes O(n) time, which is executed each `sort` iteration for every official for every match. The current implementation is therefore quite slow.

Instead of an array solution, please create an object like this:

```js
const roleOrder = {
  "Umpire": 0,
  "Reserve/Video": 1,
   ...
}
```

and access it using roleOrder[a.role]. This will drastically improve running time.

File: src/Fetchers/TMSFetcher/TMSMatchFetcher.ts [hunk 7]
```diff
@@ -46,6 +46,23 @@ export class TMSMatchFetcher {
             matches.set(item.getID(), item);
         }
 
+        // Fetch officials data
```
Comment (martijnvankekem) [inline src/Fetchers/TMSFetcher/TMSMatchFetcher.ts:49]: Instead of implementing this in the `TMSMatchFetcher.ts` class, please create a separate `TMSOfficialFetcher.ts` class and move all official-related logic into there.

File: src/Objects/Match.ts [hunk 8]
```diff
@@ -5,6 +5,12 @@ import { DateHelper } from "../Utils/DateHelper.js";
 import { Moment } from "moment-timezone";
 import { Gender, getFullGender } from "./Gender.js";
 
+export interface Official {
```
Comment (martijnvankekem) [inline src/Objects/Match.ts:8]: Please create a separate file for this.

File: src/Objects/Match.ts [hunk 9]
```diff
@@ -372,14 +384,25 @@ export class Match {
         if (this.competition)
             lines.push(`Event: ${this.competition.getName()}`);
 
-        lines.push("");
+        // Add match link if HTML is enabled
```
Comment (martijnvankekem) [inline src/Objects/Match.ts:387]: What's the advantage of this change? I don't see the added value, and it's less readable to me.

**Review by martijnvankekem — CHANGES_REQUESTED** [review]
I had some time to review the changes in-depth. 

The current implementation is much much better, awesome work! I have requested some changes though, since there still seem to be some unnecessary changes compared to `main` branch that have no added value for implementing the official-related logic.

Also, I'd like a separate class with all the official-related logic to keep it separate from the match-related logic.

**Comment by philipohagan** [comment]
Sorry for the delay, I've:
- refactored the officials logic to separate files
- reverted the other bits you wanted
- fixed the role ordering logic
